### PR TITLE
Add new datalayer value to track blocks

### PR DESCRIPTION
**Pull request by GP-Dan-Tovbein**
Ref: https://jira.greenpeace.org/browse/PLANET-7075

## Description
Create a new datalayer value to track Planet4 blocks and Gravity forms

## Testing
Create a new page with p4 or gravity form blocks and preview it.
Then, open de browser console and type `datalayer` through the terminal. Expand the `zero` index and check the value of `p4_blocks`. All the blocks should be separated by commas.

Screenshot 2023-03-21 at 16 09 38

### CODE REVIEW

File: src/Context.php [hunk 1]
```diff
@@ -95,6 +95,29 @@ public static function parse_utm_campaign_param(string $cf_local_project): strin
...
+            function ($block) {
+                if (str_contains($block, 'gravityforms')) {
+                    return substr($block, (stripos($block, ':') + 1) - strlen($block));
+                }
+
+                return substr($block, (stripos($block, '/') + 1) - strlen($block));
+            },
+            array_unique($matches[0])
+        );
+        $context['p4_blocks'] = implode(',', $p4_blocks);
```
Comment (GP-Dan-Tovbein) [inline src/Context.php:118]: Example of string `columns,gravityforms/form`

File: src/Context.php [hunk 2]
```diff
@@ -95,6 +95,31 @@ public static function parse_utm_campaign_param(string $cf_local_project): strin
...
+    public static function set_p4_blocks_datalayer(array &$context, object $post): void
+    {
+        preg_match_all('/wp:planet4-blocks\/(\S+)|wp:gravityforms\/(\S+)*/', $post->post_content, $matches);
+
+        $p4_blocks = array_map(
+            function ($block) {
+                if (str_contains($block, 'gravityforms')) {
+                    $start = stripos($block, ':');
+                    $end = stripos($block, '/');
+                    return substr($block, $start + 1, $end - strlen($block));
```
Comment (GP-Dan-Tovbein) [inline src/Context.php:113]: Take only `gravityforms` from `wp:gravityforms/form`

**Comment by mleray** [comment]
Don't we need UAT for this ticket? 🤔 Maybe Molly or Julia should check it out

**Comment by GP-Dan-Tovbein** [comment]
> Don't we need UAT for this ticket? 🤔 Maybe Molly or Julia should check it out... Either way, could you please set up a test instance for this? On local it doesn't work for me 😕

Probably because you haven't set the `Google Tag Manager Container` through `Planet 4 > Analytics`. By the way, the umbriel instance is ready :)

File: src/Context.php [hunk 3]
```diff
@@ -95,6 +95,31 @@ public static function parse_utm_campaign_param(string $cf_local_project): strin
...
 
+    /**
+     * Set p4_blocks datalayer value
+     *
+     * @param array  $context Context to be set.
+     * @param object $post That the context refers to.
+     */
+    public static function set_p4_blocks_datalayer(array &$context, object $post): void
+    {
+        preg_match_all('/wp:planet4-blocks\/(\S+)|wp:gravityforms\/(\S+)*/', $post->post_content, $matches);
```
Comment (GP-Dan-Tovbein) [inline src/Context.php:106]: This regex matches all the occurrences of `wp:planet4-blocks/` or `wp:gravityforms/` until the next empty space. 

For example, 
`wp:planet4-blocks/columns`, `wp:planet4-blocks/cover`, `wp:gravityforms/form`

**Comment by mleray** [comment]
On this [post](https://www-dev.greenpeace.org/test-umbriel/press/1111/duis-posuere-5/) there is the `Related Articles` block, but it doesn't show in the `p4_blocks`, not sure that's expected? I know it's a bit different since it's not technically part of the post content 🤔

File: src/Context.php [hunk 4]
```diff
@@ -95,6 +95,31 @@ public static function parse_utm_campaign_param(string $cf_local_project): strin
         return '';
     }
 
+    /**
+     * Set p4_blocks datalayer value
+     *
+     * @param array  $context Context to be set.
+     * @param object $post That the context refers to.
+     */
+    public static function set_p4_blocks_datalayer(array &$context, object $post): void
```
Comment (mleray) [inline src/Context.php:104]: Why is there a `&` in front of the `$context` param?

**Review by mleray — COMMENTED** [review]
Code looks good to me 👏 I only have one comment: on posts such as [this one](https://www-dev.greenpeace.org/test-umbriel/press/1111/duis-posuere-5/) there is the `Related Articles` block, but it doesn't show in the `p4_blocks`, not sure that's expected? I know it's a bit different since it's not technically part of the post content 🤔

File: single.php [hunk 5]
```diff
@@ -125,6 +125,8 @@
     ]
 );
 
+Context::set_p4_blocks_datalayer($context, $post);
```
Comment (GP-Dan-Tovbein) [inline single.php:128]: Added before the render to make sure that the `$post` has included the right content.

**Comment by GP-Dan-Tovbein** [comment]
@mleray FYI https://jira.greenpeace.org/browse/PLANET-7075?focusedId=247673&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-247673

**Review by mleray — APPROVED** [review]
Code looks good to me 👏 and it seems to work as expected, but let's wait for Molly's approval before merging 🙂

**Pull request merged**